Switch to tailwind

davidfischer opened the pull request:
This PR will switch the static assets from Bootstrap 4 to TailwindCSS 3. The actual site will still look pretty similar.

* JS dependencies are drastically reduced to just Tailwind, HTMX (used to async fetch videos and upcoming meetups)~~, and Hyperscript (used to expand/collapse the navigation menu on mobile). Hyperscript could be completely removed if we want to just have a few lines of custom JS to power the menu.~~
* ~~From a network transfer size, CSS difference (before zipping) is 143.7kb with Bootstrap vs. 7.9kb with Tailwind. JS bundle difference is 212.8kb with Bootstrap+HTMX vs. 148.5kb with HTMX+Hyperscript. Including gzipping, we're looking at 90.9kb with Bootstrap vs. 45.1kb with Tailwind for about half the total network transfer.~~ As a result, the site should load faster especially on slower network connections with these changes. The total gzipped transfer will be 20.0kb with these changes vs. 90.9kb before.

## Testing it out

* Pull and checkout this branch: `git pull && git checkout davidfischer/switch-to-tailwind`
* Install dependencies and build new static assets: `npm run install && npm run build`
* Run the dev server: `./manage.py runserver`
* Open your browser to http://localhost:8000

Note: don't forget to rebuild the old static assets when you switch to another branch.


## Screenshots
![Screenshot from 2024-06-15 09-49-59](https://github.com/sandiegopython/pythonsd-django/assets/185043/44310826-8ea9-4ebd-ac23-cb0d49657784)

![Screenshot from 2024-06-15 09-58-45](https://github.com/sandiegopython/pythonsd-django/assets/185043/6f0e8de4-8a73-4283-a85e-7ca59df7e22d)


Fixes https://github.com/sandiegopython/pythonsd-django/issues/153.

davidfischer committed:
Switching to tailwind

davidfischer committed:
Merge branch 'main' into davidfischer/switch-to-tailwind

davidfischer committed:
Update remaining templates to TailwindCSS

- Remove Webpack (woot!)
- Add Hyperscript for the navigation menu

davidfischer committed:
Build static assets for production in Docker

Jer-Pha commented on package.json:
```diff
@@ -4,15 +4,16 @@
   "description": "",
   "main": "index.js",
```
`index.js` is being deleted. Is this line still necessary?

I'm still learning npm; this is definitely a question, not a correction.

davidfischer replied:
I thought it might be required, but I [double-checked](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#main) and it is not. Removed

davidfischer committed:
Update package.json

- Rev the minor version
- Add a description
- Update GitHub links

davidfischer committed:
Remove Hyperscript from Tailwind migration

- Menu toggle is handled by custom JS

davidfischer commented:
I also added a commit to remove Hyperscript. It seems like overkill and an unnecessary dependency for now. If it becomes useful, we'll add it back but seems odd to add a huge dependency instead of 3 lines of JS.

davidfischer committed:
Slight refactor of custom JS

- Move under the pythonsd/ directory where the template that uses it
  lives
- Remove copying static files in package.json.
  Not needed after moving the main.js
- Remove logging when debug false (can't get errors otherwise)

davidfischer committed:
Put menu "first" but make it async

Jer-Pha commented on package-lock.json:
```diff
@@ -9,189 +9,129 @@
-        "jquery": "^3.6.0",
-        "popper.js": "^1.14.3"
+        "htmx.org": "^1.9.12",
+        "hyperscript.org": "^0.9.12"
```
```suggestion
```

davidfischer replied:
Thanks for pointing that out. I hadn't re-run the install step after removing it. This file is automatically generated so best to let the generator do its thing.

davidfischer committed:
Update package-lock

Jer-Pha commented on pythonsd/static/js/main.js:
```diff
@@ -0,0 +1,5 @@
+// Toggle the navigation menu on small screens
+document.getElementById("navigation-toggle").addEventListener("click", function () {
+    const navMenu = document.querySelector("#navigation-menu");
```
```suggestion
    const navMenu = document.getElementById("navigation-menu");
```
getElementById() is faster than querySelector(), [source](https://stackoverflow.com/a/71878632). Obviously the performance difference won't be noticeable, I just like this optimization.

Personally, I would also combine lines 3 and 4 and not bother with assigning `navMenu`. Less readable, though, so I'm not suggesting to change it.

davidfischer replied:
Looks good to me. Thanks.

davidfischer committed:
Update pythonsd/static/js/main.js

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

Jer-Pha commented:
This is more of a general conversation but it could be grouped into these changes:

Do we want `target="_blank"` for links that leave the site? It's currently on the MeetUp links on the organizer page (from my original PR), but no other external links use it. It comes down to [Usability vs. Marketing](https://stackoverflow.com/a/1349067). Even though I would personally add it to all external links, I think the target audience of this site doesn't need that choice made for them.

Jer-Pha commented on pythonsd/templates/pythonsd/index.html:
```diff
@@ -7,84 +7,81 @@
+      <div>
+        <h2>About us</h2>
+        <p>San Diego Python is a group for those interested in learning about Python or connecting with other Python professionals. We have a smaller weekly meetup for people to work on Python projects and a larger monthly meetup on the 4th Thursday of the month with speakers and talks.</p>
+        <p>Join us on <a href="https://www.meetup.com/pythonsd/">Meetup</a> to stay up to date with our upcoming events!</p>
```
```suggestion
        <p>Join us on <a href="https://www.meetup.com/pythonsd/" rel="nofollow noopener noreferrer">Meetup</a> to stay up to date with our upcoming events!</p>
```

Jer-Pha commented on pythonsd/templates/pythonsd/index.html:
```diff
@@ -7,84 +7,81 @@
+          <img src="{% static 'img/sponsors/neo4j-logo.png' %}" class="flex-none h-16" alt="Neo4j">
+          <figcaption class="flex-auto">
+            <h5 class="mt-0 mb-1"><a href="https://neo4j.com/" rel="nofollow noopener noreferrer">Neo4j</a></h5>
+            <p class="m-0">Neo4j is providing food at our monthly meetup</p>
```
```suggestion
            <p class="m-0">Neo4j is providing food at our monthly meetup.</p>
```

davidfischer committed:
Update pythonsd/templates/pythonsd/index.html

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

davidfischer committed:
Update pythonsd/templates/pythonsd/index.html

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

davidfischer commented:
> Do we want `target="_blank"` for links that leave the site?

I don't feel particularly strong but I generally don't open links in new windows unless there's a reason or an expectation that that should happen.

Jer-Pha commented on pythonsd/templates/pythonsd/organizers.html:
```diff
@@ -5,39 +5,40 @@
-          <li class="list-inline-item">
-            <a href="{{ organizer.meetup_url }}" rel="nofollow noopener noreferrer" target="_blank">
+          <div class="w-6 h-6">
+            <a class="text-slate-500 hover:text-slate-800" href="{{ organizer.meetup_url }}" rel="nofollow noopener noreferrer" target="_blank">
```
```suggestion
            <a class="text-slate-500 hover:text-slate-800" href="{{ organizer.meetup_url }}" rel="nofollow noopener noreferrer">
```

Jer-Pha commented on pythonsd/templates/pythonsd/organizers.html:
```diff
@@ -5,39 +5,40 @@
-          <li class="list-inline-item">
-            <a href="{{ organizer.linkedin_url }}" rel="nofollow noopener noreferrer" target="_blank">
+          <div class="w-6 h-6">
+            <a class="text-slate-500 hover:text-slate-800" href="{{ organizer.linkedin_url }}" rel="nofollow noopener noreferrer" target="_blank">
```
```suggestion
            <a class="text-slate-500 hover:text-slate-800" href="{{ organizer.linkedin_url }}" rel="nofollow noopener noreferrer">
```

Jer-Pha reviewed (changes requested):
Only minor suggestions based on the external link conversation. The switch to Tailwind looks good. Good call removing Hyperscript for a single element.

davidfischer committed:
Update pythonsd/templates/pythonsd/organizers.html

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

davidfischer committed:
Update pythonsd/templates/pythonsd/organizers.html

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

Jer-Pha commented on pythonsd/templates/pythonsd/organizers.html:
```diff
@@ -5,39 +5,40 @@
-        <h5 class="card-title">{{ organizer.name }}</h5>
-        <ul class="list-inline">
+    <div class="flex flex-col rounded border border-slate-200 max-w-64 md:max-w-80 mx-auto">
+      <img src="{{ organizer.photo.url }}" class="flex-none" alt="{{ organizer.name }}">
```
```suggestion
      <img src="{{ organizer.photo.url }}" class="flex-none aspect-square object-cover" alt="{{ organizer.name }}">
```
This preserves the aspect ratio if the uploaded file is not 1:1.

davidfischer replied:
Definitely a defensive improvement. Thanks.

Jer-Pha commented on pythonsd/templates/pythonsd/organizers.html:
```diff
@@ -5,39 +5,40 @@
 
 {% if organizers %}
-<div class="row row-cols-2 row-cols-md-4">
+<div class="grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-12 mb-16 ">
```
```suggestion
<div class="grid grid-cols-1 sm:grid-cols-2  md:grid-cols-3 lg:grid-cols-4 gap-12 mb-16 ">
```
640-767px isn't the most common viewport width, but this will remove unnecessary white space for users with that resolution.

Jer-Pha commented on pythonsd/templates/pythonsd/base.html:
```diff
@@ -22,99 +21,93 @@
+        </a>
+
+        <!-- Navigation menu toggle -->
+        <div id="navigation-toggle" class="h-6 w-6 cursor-pointer md:hidden block text-slate-100">
+          <svg xmlns="http://www.w3.org/2000/svg" id="menu-button" fill="none" viewBox="0 0 24 24" stroke="currentColor">
+            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h16"></path>
+          </svg>
+        </div>
+
+        <div id="navigation-menu" class="hidden w-full mt-4 md:w-auto md:flex md:items-center md:mt-0">
+          <ul class="md:flex md:justify-between font-extrabold mb-0">
+            <li class="py-2 md:mr-5">
+              {% url 'code-of-conduct' as conduct_url %}
+              <a class="text-slate-100 hover:text-slate-300" href="{% if request.path != conduct_url %}{{ conduct_url }}{% else %}#{% endif %}">Code of Conduct</a>
+            </li>
+            <li class="py-2 md:mr-5">
```
```suggestion
        <div id="navigation-toggle" class="h-6 w-6 cursor-pointer sm:hidden block text-slate-100">
          <svg xmlns="http://www.w3.org/2000/svg" id="menu-button" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h16"></path>
          </svg>
        </div>

        <div id="navigation-menu" class="hidden w-full mt-4 sm:w-auto sm:flex sm:items-center sm:mt-0">
          <ul class="sm:flex sm:justify-between font-extrabold mb-0">
            <li class="py-2 sm:mr-5">
              {% url 'code-of-conduct' as conduct_url %}
              <a class="text-slate-100 hover:text-slate-300" href="{% if request.path != conduct_url %}{{ conduct_url }}{% else %}#{% endif %}">Code of Conduct</a>
            </li>
            <li class="py-2 sm:mr-5">
```
Ignore this if there are plans to add a third link.

davidfischer replied:
I think I'm going to skip this change. There aren't plans to add a 3rd top navigation link but I also want to keep it simple if we do decide to change it. I'm also not particularly worried if users with screens narrower than 768px have to toggle the nav button.

Jer-Pha commented on pythonsd/templates/pythonsd/index.html:
```diff
@@ -7,84 +7,81 @@
+      <div>
+        <h2>Latest video</h2>
+
+        <div id="youtube-widget" hx-get="{% url 'recent_videos' %}" hx-trigger="load">
```
The two `hx-get` requests could be combined into a single request using [hx-swap-oob](https://htmx.org/attributes/hx-swap-oob/). This would require a third view that calls the other two. Probably not necessary unless there's concern over bandwidth.

davidfischer replied:
I am not particularly concerned about bandwidth and the [overhead of a 2nd request to the same domain in HTTP2 (used in prod) is pretty minimal](https://www.cloudflare.com/learning/performance/http2-vs-http1.1/). For now, I think the current setup is simpler, allows the views to fail/succeed independently, and easily allows them to be cached separately.

Jer-Pha commented on pythonsd/templates/pythonsd/base.html:
```diff
@@ -22,99 +21,93 @@
 
     {# FOOTER #}
-    <footer class="container">
+    <footer class="container mx-auto">
```
(Personal preference) I liked the old footer better on larger displays (more compact) but the new format looks better on mobile. (stacked lines and centered icons).

davidfischer replied:
This is fair. I think I like the old footer as well on larger displays although I'm not particularly worried about how compact it is. I think I like the look of the new icon spacing a bit better in the new footer but I don't love the spacing of text in the new footer.

I wonder if we could switch it to 2 text sections instead of 3. In reality, "edit on github" and the version could be condensed into one as they are related. See screenshots. Thoughts?

![Screenshot from 2024-06-22 07-46-33](https://github.com/sandiegopython/pythonsd-django/assets/185043/74923bcf-9adf-4603-adc4-fb77fb659665)
![Screenshot from 2024-06-22 07-46-53](https://github.com/sandiegopython/pythonsd-django/assets/185043/e5ef3769-b134-4691-8952-0867176e80c3)

davidfischer replied:
I merged this PR without changing this but I'd be happy to do a separate PR with this if you think this is an improvement.

Jer-Pha reviewed (approved):
Overall, this looks great! None of these suggestions are necessary. Just minor formatting stuff. This looks ready to merge as-is.

davidfischer committed:
Update pythonsd/templates/pythonsd/organizers.html

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

davidfischer committed:
Update pythonsd/templates/pythonsd/organizers.html

Co-authored-by: Jeremy <131234246+Jer-Pha@users.noreply.github.com>

willingc reviewed (approved):
Love this! Thanks @davidfischer for simplifying and getting rid of webpack :wink:
